### PR TITLE
elastic search

### PR DESCRIPTION
**Pull request by EngrEric**
This branch adds elastic search feature

### CODE REVIEW

File: esClient.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import elasticSearch from 'elasticsearch';
+
+export default new elasticSearch.Client({
+    host: 'http://localhost:9200',
```
Comment (enyeInc) [inline esClient.js]: this should be an env variable

File: src/components/home/components/Search.js [hunk 2]
```diff
@@ -1,22 +1,76 @@
...
+                        <Select
+                            className="select-category"
+                            placeholder="Select Category"
+                            onChange={value => this.handleChange(value)}
+                        >
+                            <Option value="name">{'Product Name'}</Option>
+                            <Option value="quantity">{'Quantity'}</Option>
+                            <Option value="cost">{'Cost'}</Option>
+                            <Option value="type">{'Type'}</Option>
+                            <Option value="description">{'Description'}</Option>
```
Comment (enyeInc) [inline src/components/home/components/Search.js]: map over an array

File: src/components/productList/components/ProductList.js [hunk 3]
```diff
@@ -1,25 +1,72 @@
...
+
+        this.setState({ productList: json.data });
+    }
+
+    // componentDidUpdate(){
+    //     const { userSearchedProducts } = this.props;
+    //     if(!userSearchedProducts.length) {
+    //         userSearchedProducts[0].foundProducts.length > 0 && this.setState();
+    //     }
+    // }
```
Comment (enyeInc) [inline src/components/productList/components/ProductList.js]: clean this up

File: src/components/productList/components/ProductList.js [hunk 4]
```diff
@@ -1,25 +1,72 @@
...
+
     render() {
+        const { productList } = this.state;
         const { path } = this.props;
-
         return (
             <React.Fragment>
                 {path !== '/home' && <CarouselImages />}
+                {/*   {this.state.noProductsFromSearch ?
+                    <Empty  /> : */}
```
Comment (enyeInc) [inline src/components/productList/components/ProductList.js]: clean up code

File: db/models/product.js [hunk 5]
```diff
@@ -1,19 +1,31 @@
 import mongoose from 'mongoose';
+import mongoosastic from 'mongoosastic';
+
+import esClient from '../../esClient';
+//import User from './user';
```
Comment (enyeInc) [inline db/models/product.js]: clean up

File: routes/product/helpers.js [hunk 6]
```diff
@@ -19,6 +19,23 @@ export default {
             return res.json({ error, success: false });
         }
     },
+    //clean up invalid products and producers
+    cleanProductsWithInvalidProducerId: async (req, res) => {
+        try {
+            const data = await Product.find();
+            const invalidProducer = data.map(dataitem => {
+                if (!ObjectID.isValid(dataitem.producerId) || !ObjectID.isValid(dataitem._id)) {
```
Comment (enyeInc) [inline routes/product/helpers.js:27]: why are you doing this check here?

File: routes/search/constant.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+    'email',
+    'firstName',
+    'lastName',
+    'phoneNumber',
+    'state',
+    'address',
+    'country',
+    'address',
+    'city',
+];
```
Comment (enyeInc) [inline routes/search/constant.js]: this is repeated in `routes/product/constants.js`

File: routes/search/searchProducts.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+                { match: req.query },
+                { stored_fields: ['_id'] },
+            );
+            const esFoundProductIds = products.hits.hits.map(({ _id }) => _id);
+            const foundProducts = await Product.find({ _id: { $in: esFoundProductIds } }).populate(
+                'producerId',
+                PRODUCER_DATA,
+            );
+
+            return res.status(OK).send({ foundProducts, numOfFoundProducts: products.hits.total });
```
Comment (enyeInc) [inline routes/search/searchProducts.js]: make sure that this matches Jsend conventions (https://github.com/omniti-labs/jsend)

File: src/components/home/actions.js [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+
+export const searchQuery = payload => ({
+    payload,
+    type: REQUEST_SEARCH,
+});
+
+export const updateSearchResult = payload => ({
+    payload,
+    type: UPDATE_SEARCH,
+});
```
Comment (enyeInc) [inline src/components/home/actions.js:25]: document your actions

File: src/components/home/components/Search.js [hunk 10]
```diff
@@ -1,22 +1,76 @@
...
+import { connect } from 'react-redux';
+import PropTypes from 'prop-types';
 
 import { SEARCH } from '../constants';
+import * as searchActions from '../actions';
+import { getSearchProducts } from '../selectors';
 
+const { Option } = Select;
 const Search = Input.Search;
+const InputGroup = Input.Group;
```
Comment (enyeInc) [inline src/components/home/components/Search.js:14]: be consistent with component variable declarations look at `line 11 - 13`

File: src/components/home/components/Search.js [hunk 11]
```diff
@@ -1,22 +1,76 @@
...
+                            onChange={value => this.handleChange(value)}
+                        >
+                            <Option value="name">{'Product Name'}</Option>
+                            <Option value="quantity">{'Quantity'}</Option>
+                            <Option value="cost">{'Cost'}</Option>
+                            <Option value="type">{'Type'}</Option>
+                            <Option value="description">{'Description'}</Option>
+                        </Select>
+                        <Search
+                            style={{ width: '70%' }}
```
Comment (enyeInc) [inline src/components/home/components/Search.js]: move styling to css sheet

File: src/components/home/components/Search.js [hunk 12]
```diff
@@ -1,22 +1,76 @@
...
+                            placeholder={SEARCH_PRODUCTS}
+                            onSearch={value =>
+                                searchQuery({ searchTerm: this.state.selectedTerm, value })
+                            }
```
Comment (enyeInc) [inline src/components/home/components/Search.js]: this should be  `onSearch= handleSearchQuery`
```
handleSearchQuery = value => {
        const { searchQuery } = this.props.actions;
        const { selectedTerm } = this.state;
       
       searchQuery({ selectedTerm, value});
}
```
more efficient, this way a new `onSearch` function isnt being created whenever this component is re-rendered

File: src/components/home/components/Search.js [hunk 13]
```diff
@@ -1,22 +1,76 @@
...
     }
 }
 
-export default SearchItems;
+SearchItems.propTypes = {
+    actions: PropTypes.object,
+};
+
+const mapDispatchToProps = dispatch => ({
+    actions: bindActionCreators(searchActions, dispatch),
```
Comment (enyeInc) [inline src/components/home/components/Search.js]: you probably dont want to bind all the actions to this component. Seems that you are only using `searchQuery`

File: src/components/home/sagas.js [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+import { updateSearchResult } from './actions';
+
+function* searchProduct(action) {
+    try {
+        const {
+            payload: { searchTerm, value },
+        } = action;
+
+        const response = yield fetch(`${SEARCH_URL}?${searchTerm}=${value}`, {
+            method: 'GET',
```
Comment (enyeInc) [inline src/components/home/sagas.js]: i dont think you have to specific `GET` here

File: src/components/home/selectors.js [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * Selects the <tt>search string</tt> key.
+ *
+ * @function
+ * @param {Object} state - redux store state
+ * @return {Object} the search result state of the home
+ * {@link module:home/constants::INITIAL_STATE constants::INITIAL_STATE}).
+ */
+export const getUserSearchParams = state => state[NAME].payload;
```
Comment (enyeInc) [inline src/components/home/selectors.js]: why do you have a `payload` key on this piece of state?

File: src/components/home/reducers.js [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+};
+
+export default (state = { ...initialState }, action) => {
+    switch (action.type) {
+        case REQUEST_SEARCH: {
+            const { payload } = action;
+            return {
+                ...state,
+                isloading: true,
+                payload,
```
Comment (enyeInc) [inline src/components/home/reducers.js]: no this is wrong, there should not be a `payload` key on the state. What is payload?

File: src/components/productList/components/ProductList.js [hunk 17]
```diff
@@ -1,25 +1,72 @@
...
+        const { productList } = prevState;
+
+        if (userSearchedProducts.length) {
+            const { foundProducts, numOfFoundProducts } = userSearchedProducts[0];
+
+            if (foundProducts.length > 0) {
+                const productList = foundProducts;
+
+                return (
+                    message.info(`We found ${numOfFoundProducts} results for your search`, 3) && {
```
Comment (enyeInc) [inline src/components/productList/components/ProductList.js]: why are you doing the messaging in a return statement?

File: routes/product/helpers.js (hunk 6 above)
Comment (EngrEric) [reply routes/product/helpers.js:27]: We added some invalid producers and products previously when we started. their IDs are assigned manually then. Here, i am using this check to find them and delete them.
it supposed to be a one time feature but i allowed it for redundancy purposes

File: src/components/home/reducers.js (hunk 16 above)
Comment (EngrEric) [reply src/components/home/reducers.js]: i don't even need this piece of the state :)

**Pull request merged**